llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -1362,6 +1364,10 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
         for (MVT Ty : {MVT::v4f16, MVT::v8f16})
           setOperationAction(Op, Ty, Legal);
     }
+    if (!Subtarget->hasFullFP16()) {
+      setOperationPromotedToType(ISD::FCANONICALIZE, MVT::v4f16, MVT::v4f32);
```

Comment:
Can these be moved close to the existing setOperationPromotedToType around line 810?